lib/include/jxl/cms_interface.h

```diff
@@ -0,0 +1,210 @@
+/* Copyright (c) the JPEG XL Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** @file cms_interface.h
+ * @brief Interface to allow the injection of different color management systems
+ * (also called color management modules, or CMMs) in JPEG XL.
+ *
+ * A CMS is needed by the JPEG XL encoder and decoder to perform colorspace
+ * conversions. This defines an interface that can be implemented for different
+ * CMSes and then passed to the library.
+ */
+
+#ifndef JXL_CMS_INTERFACE_H_
+#define JXL_CMS_INTERFACE_H_
+
+#include "jxl/color_encoding.h"
+#include "jxl/memory_manager.h"
+
+#if defined(__cplusplus) || defined(c_plusplus)
+extern "C" {
+#endif
+
+/** Represents an input or output colorspace to a color transform, as a
+ * serialized ICC profile. */
+typedef struct {
+  /** The serialized ICC profile. */
+  struct {
+    const uint8_t* data;
+    size_t size;
+  } icc;
+
+  /** Structured representation of the colorspace, if applicable. */
+  JxlColorEncoding color_encoding;
+
+  /** Number of components per pixel. */
+  size_t num_channels;
+} JxlColorProfile;
+
+/** Allocates and returns the data needed for @p num_threads parallel transforms
+ * from the @p input colorspace to @p output, with up to @p pixels_per_thread
+ * pixels to transform per call to JxlCmsInterface::run. @p init_data comes
+ * directly from the JxlCmsInterface instance. Since @c run only receives the
+ * data returned by @c init, a reference to @p init_data should be kept there
+ * if access to it is desired in @c run. Likewise for JxlCmsInterface::destroy.
+ *
+ * The ICC data in @p input and @p output is guaranteed to outlive the @c init /
+ * @c run / @c destroy cycle.
+ *
+ * @param init_data JxlCmsInterface::init_data passed as-is.
+ * @param num_threads the maximum number of threads from which
+ *        JxlCmsInterface::run will be called.
+ * @param pixels_per_thread the maximum number of pixels that each call to
+ *        JxlCmsInterface::run will have to transform.
+ * @param input the input colorspace for the transform.
+ * @param output the colorspace to which JxlCmsInterface::run should convert the
+ *        input data.
+ * @param intensity_target for colorspaces where luminance is relative
+ *        (essentially: not PQ), indicates the luminance of (1, 1, 1).
```

Comment:
How is `intensity_target` supposed to be interpreted / taken into account by a CMS? Does it apply to the input or output colorspace?

Reply:
It is only relevant if one of the profiles is relative and the other is absolute, in which case it applies to relative one, whether that is the input or the output. Do you think this should be made somehow clearer?

Reply:
If this is a parameter that gets ignored in the case of absolute to absolute or relative to relative, that should be documented.

I think @lvandeve's suggestion might make sense to have this field in JxlColorProfile, maybe with a special value for "relative". Conceptually it makes more sense to me to have "fully defined"  colorspaces (not necessarily actually fully defined in the case of relative luminance, but at least all available info is in the struct that way).

Reply:
The input intensity target matters if the input is in a relative colorspace and the output is absolute, for example (1, 1, 1) in sRGB would be converted to (0.6046, 0.6046, 0.6046) in PQ if `intensity_target = 255`, or to (0.8274, 0.8274, 0.8274) if `intensity_target = 2000`.

Conversely, (0.6046, 0.6046, 0.6046) in PQ would be converted to (1, 1, 1) in sRGB if `intensity_target = 255`, or to (0.3922, 0.3922, 0.3922) if `intensity_target = 2000`.

Regarding making it part of the colorspace vs. the image, the way I see it is:

1) currently, in the rest of the codebase, we consider it a property of the image. It is the maximum luminance that can possibly be found in the image, and corresponds to (1, 1, 1) when the image is encoded using a relative colorspace. So there is some logic in having it that way, and it also makes it clear that relative<->relative conversions won’t need to concern themselves about this as it will be the same on both sides.

2) Alternatively, we could consider it a property of the colorspace, but then it would deviate from the meaning that the term “intensity_target” currently has so we should probably change it in this interface to make that clear, for example to “white_luminance” as in EXR. The invoker of the CMS would then set that to `intensity_target` if the colorspace is relative, or to 10 000 for PQ. Then, there would be less need to treat PQ specially in the CMS: the condition for the luminance rescaling would instead be `input_white_luminance ≠ output_white_luminance`. So there is indeed logic in support of this option as well although it would require more changes to the current code.

Reply:
Option 2 seems most attractive to me.

In general, it would also be good to clarify the semantics of intensity_target in the bitstream, in the encoder, the decoder, the butteraugli metric, and here. It's indeed a good idea to use different names when it means different things (or has different implications).

Reply:
It should be clear to the user how to use the field or parameter, the user has to implement this function, and e.g. if you implement it using lcms or skcms, it'd be good if the intensity_target parameter is well documented (or alternatively there is a good example showing how to use it) so that the user knows if they have to use the parameter before or after calling the CMS, or it's a setting the CMS itself supports

Reply:
I would favor sticking with 1, at least for now. I think it brings fewer unnecessary complications. I have added some clarification of how to interpret the `intensity_target` parameter, how does it look?

Reply:
Given the meaning of "brightest luminance present in the image", having one seems fine